examples/single_mem_node/main.rs

```diff
@@ -0,0 +1,198 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate raft;
+
+use std::sync::mpsc::{self, RecvTimeoutError};
+use std::time::{Duration, Instant};
+use std::thread;
+use std::collections::HashMap;
+
+use raft::prelude::*;
+use raft::storage::MemStorage;
+
+type ProposeCallback = Box<Fn() + Send>;
+
+enum Msg {
+    Propose { id: u8, cb: ProposeCallback },
+    #[allow(dead_code)] Raft(Message),
+}
+
+// A simple example about how to use the raft library in Rust.
+fn main() {
+    // Create a storage for Raft, here we just use a simple memory storage.
+    // You need to build your own persistent storage in your production.
+    // Please check the Storage traint in src/storage.rs to see how to implement one.
+    let storage = MemStorage::new();
+
+    // Create the configuration for the Raft node.
+    let cfg = Config {
+        // The unique ID for the Raft node.
+        id: 1,
+        // The Raft node list.
+        // Mostly, the peers need to be saved in the storage
+        // and we can get them from Storage::initial_state function, so here
+        // you need to set empty.
+        peers: vec![1],
+        // Election tick is for how long the follower may campaign again after
+        // it doesn't receive any message from the leader.
+        election_tick: 10,
+        // Heartbeat tick is for how long the leader needs to send
+        // a heartbeat to keepalive.
+        heartbeat_tick: 3,
+        // The max size for one message, mostly, 1 MB is enough.
+        max_size_per_msg: 1024 * 1024 * 1024,
+        // Max inflight msgs that the leader sends message to follower without
+        // receiving ACKs.
+        max_inflight_msgs: 256,
+        // The Raft applied index.
+        // You need to save your applied index when you apply the committed Raft logs.
+        applied: 0,
+        // Just for log
+        tag: format!("[{}]", 1),
+        ..Default::default()
+    };
+
+    // Create the Raft node.
+    let mut r = RawNode::new(&cfg, storage, vec![]).unwrap();
+
+    let (sender, receiver) = mpsc::channel();
+
+    // Use another thread to propose a Raft request.
+    send_propose(sender);
+
+    // Loop forever to drive the Raft.
+    let mut t = Instant::now();
+    let d = Duration::from_millis(100);
+
+    // Use a HashMap to holde the propose callbacks.
+    let mut cbs = HashMap::new();
+
+    loop {
+        match receiver.recv_timeout(d) {
+            Ok(Msg::Propose { id, cb }) => {
+                cbs.insert(id, cb);
+                r.propose(vec![id], false).unwrap();
+            }
+            Ok(Msg::Raft(m)) => r.step(m).unwrap(),
+            Err(RecvTimeoutError::Timeout) => (),
+            Err(RecvTimeoutError::Disconnected) => return,
+        }
+
+        if t.elapsed() >= d {
```

Comment:
`d` should be updated if `elapsed` is less than d.

Reply:
here I don't need an accurate check, and the elapsed is based on the created time, so the elapsed time will be increased next time, seem no need to update d.

Reply:
For example, if a message arrive right before timeout, then next tick is called after `d * 2` instead of d.